lib/fs/dumbfs.c

```diff
@@ -42,32 +43,125 @@ dumbfs_unmount(void *cookie)
-	char const *path __unused /* TODO */,
-	struct filehandle *handle __unused /* TODO */
+	struct file_handle *handle,
+	char const *path
 )
 {
-	/* TODO */
-	return (ERR_NOT_FOUND);
+	uint32 file_index;
+	struct dumbfs_file_entry file_header[1];
+	struct fs_mount *mount;
+	struct fs_dumb *dumb;
+	uint offset;
+	size_t path_len;
+	char *file_path;
+	status_t err;
+	struct dumbfs_file *file;
+
+	err = OK;
+	mount = handle->mount;
+	dumb = mount->fs_data;
+	if (!path[0]) {
+		handle->type |= FS_DIRECTORY;
+		handle->file_data = NULL;
+		return (OK);
+	}
+	path_len = strlen(path);
+	if (!(file_path = kalloc(path_len + 1))) {
+		return (ERR_NO_MEMORY);
+	}
+	file_index = 0;
+	offset = sizeof(uint32_t);
+	while (file_index < dumb->nb_files) {
+		if (bdev_read(mount->device, file_header, offset, sizeof(*file_header)) != sizeof(*file_header)
+			|| bdev_read(mount->device, file_path, offset + sizeof(*file_header), path_len + 1) != (ssize_t)path_len + 1)
+		{
+			err = ERR_BAD_DEVICE;
+			goto end;
+		}
+		offset += sizeof(struct dumbfs_file_entry);
+		if (!strncmp(path, file_path, path_len + 1)) {
+			if (!(file = kalloc(sizeof(*file)))) {
+				err = ERR_NO_MEMORY;
+				goto end;
+			}
+			memcpy(&file->header, file_header, sizeof(file->header));
+			file->bdev_offset = offset + path_len + 1;
+			file->seek_offset = 0;
+			handle->file_data = file;
+			goto end;
+		}
+		offset += file_header->entry_size;
+		++file_index;
+	}
+	err = ERR_NOT_FOUND;
+end:
+	kfree(file_path);
+	return (err);
+}
+
+static status_t
+dumbfs_opendir(
+	struct dir_handle *dir_handle
+)
+{
+	struct dumbfs_dir *dir;
+
+	if (!(dir = kalloc(sizeof(*dir)))) {
+		return (ERR_NO_MEMORY);
+	}
+	dir->bdev_offset = sizeof(uint);
+	dir->file_index = 0;
+	dir_handle->dir_data = dir;
+	return (OK);
 }
 
 static status_t
 dumbfs_close(
-	void *fscookie __unused /* TODO */,
-	struct filehandle *handle __unused /* TODO */
+	struct file_handle *handle
 )
 {
+	kfree(handle->file_data);
 	return (OK);
 }
 
 static status_t
 dumbfs_readdir(
-	void *fscookie __unused /* TODO */,
-	void *data __unused /* TODO */,
-	struct dirent *dirent __unused /* TODO */
+	struct dir_handle *dir_handle,
+	struct dirent *dirent
+)
+{
+	struct dumbfs_dir *dir;
+	struct fs_mount *mount;
+	struct fs_dumb *dumb;
+	struct dumbfs_file_entry file_header[1];
+	ssize_t r;
+
+	mount = dir_handle->file_handle->mount;
+	dir = dir_handle->dir_data;
+	dumb = mount->fs_data;
+	if (dir->file_index == dumb->nb_files) {
+		return (ERR_DIRECTORY_END);
+	}
+	dirent->dir = false;
+	if (bdev_read(mount->device, file_header, dir->bdev_offset, sizeof(*file_header)) != sizeof(*file_header)) {
+		return (ERR_BAD_DEVICE);
+	}
+	r = bdev_read(mount->device, file_header, dir->bdev_offset + sizeof(*file_header), sizeof(dirent->name) - 1);
+	if (r == -1) {
+		return (ERR_BAD_DEVICE);
+	}
+	dir->bdev_offset += sizeof(struct dumbfs_file_entry) + file_header->entry_size;
+	++dir->file_index;
+	dirent->name[r] = 0;
```

Comment:
Am i mistaken or you are never filling the `dirent` structure?

Reply:
typo above, I am reading twice into file_header. Will fix